### PR TITLE
Fetching data from API, BLoC implementation and unit and widget tests.

### PR DESCRIPTION
**Pull request by FacuCristobalGlobant**
This time I used API calls to fetch the card's data and implemented the BLoC pattern to manage the busines logic. Then, I made some unit and widget tests for my app. Also changed the ListView widget for a GridView and implemented a Drawer.

https://user-images.githubusercontent.com/109153713/187925909-cfd51553-4dee-47d3-9aa3-855838ab3e94.mp4

The following image shows all the tests I implemented executing without errors.

Screenshot 2022-09-05 120052

### CODE REVIEW

File: lib/src/core/util/api_response_state.dart [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import 'package:hearthstoneapp/src/data/model/hearthstone_card.dart';
+
+class ApiResponse {
+  final List<HearthstoneCard>? listOfCards;
+  final ApiResponseState state;
+
+  ApiResponse({required this.state, this.listOfCards});
```
Comment (FranciscoGaraventa) [inline lib/src/core/util/api_response_state.dart]: trailing comma

File: lib/src/data/datasource/remote/api_service.dart [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import 'package:hearthstoneapp/src/core/util/constants.dart';
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/remote/api_service.dart]: relative import

File: lib/src/data/repository/card_list_repository.dart [hunk 3]
```diff
@@ -1,20 +1,33 @@
...
+          final HearthstoneCard currentCard =
+              HearthstoneCard.fromJson(card as Map<String, dynamic>);
+          listOfCards.add(currentCard);
+        });
+      }
+    } catch (e) {
+      throw Exception('there was an error retrieving data');
+    }
 
     return listOfCards;
```
Comment (FranciscoGaraventa) [inline lib/src/data/repository/card_list_repository.dart:28]: In the case of returning an empty list, does that ensure that you had an error in the response?

File: lib/src/domain/bloc/i_hearthstone_card_list_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import 'package:hearthstoneapp/src/core/util/api_response_state.dart';
+import 'package:hearthstoneapp/src/data/model/hearthstone_card.dart';
+import 'package:hearthstoneapp/src/domain/bloc/i_bloc.dart';
```
Comment (FranciscoGaraventa) [inline lib/src/domain/bloc/i_hearthstone_card_list_bloc.dart:3]: relative imports?

File: lib/src/presentation/bloc/list_of_cards_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import 'dart:async';
+import 'package:hearthstoneapp/src/core/util/api_response_state.dart';
+import 'package:hearthstoneapp/src/data/model/hearthstone_card.dart';
+import 'package:hearthstoneapp/src/data/repository/card_list_repository.dart';
+import 'package:hearthstoneapp/src/domain/bloc/i_hearthstone_card_list_bloc.dart';
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/list_of_cards_bloc.dart:5]: relative imports?

File: lib/src/presentation/widget/card_property.dart [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import 'package:flutter/material.dart';
 import 'package:hearthstoneapp/src/core/util/constants.dart';
 import 'package:hearthstoneapp/src/core/util/custom_flex.dart';
+import 'package:hearthstoneapp/src/core/util/keys.dart';
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/card_property.dart]: relative imports?

File: lib/src/presentation/widget/not_found.dart [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    return Center(
+      child: SingleChildScrollView(
+        child: Column(
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [
+            Image.asset(Constants.notFoundImage),
+            const Text(
+              Constants.notFoundMessage,
+              style: TextStyle(
+                fontSize: 25,
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/not_found.dart]: add this value to const

File: lib/src/presentation/widget/search_menu.dart [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: [
+        ColoredBox(
+          color: Palette.boxColor,
+          child: Padding(
+            padding: const EdgeInsets.symmetric(vertical: 30),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Add this value to const

File: lib/src/presentation/widget/search_menu.dart [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            child: Column(
+              children: [
+                Image.asset(
+                  Constants.hearthstoneLogo,
+                  height: 80,
+                ),
+                Container(
+                  margin: const EdgeInsets.only(top: 20, left: 20, right: 20),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Same as above

File: lib/src/presentation/widget/search_menu.dart [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      children: [
+        ColoredBox(
+          color: Palette.boxColor,
+          child: Padding(
+            padding: const EdgeInsets.symmetric(vertical: 30),
+            child: Column(
+              children: [
+                Image.asset(
+                  Constants.hearthstoneLogo,
+                  height: 80,
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Same as above

File: lib/src/presentation/widget/search_menu.dart [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                      keyword = value;
+                    },
+                  ),
+                )
+              ],
+            ),
+          ),
+        ),
+        Padding(
+          padding: const EdgeInsets.symmetric(vertical: 15.0, horizontal: 20),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Same as above

File: lib/src/presentation/widget/search_menu.dart [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+          child: Column(
+            mainAxisAlignment: MainAxisAlignment.end,
+            crossAxisAlignment: CrossAxisAlignment.stretch,
+            children: [
+              DrawerButton(
+                onClickAction: () {
+                  widget.hearthstoneCardListBloc
+                      .searchForCards(Endpoints.byName + keyword);
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Instead to do this, you can use string interpolation '${Endpoints.byName}$keyword'

File: lib/src/presentation/widget/search_menu.dart [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+                  Scaffold.of(context).closeDrawer();
+                },
+                text: Constants.searchByNameText,
+              ),
+              DrawerButton(
+                onClickAction: () {
+                  widget.hearthstoneCardListBloc
+                      .searchForCards(Endpoints.byFaction + keyword);
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Same as above

File: lib/src/presentation/widget/search_menu.dart [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+                  Scaffold.of(context).closeDrawer();
+                },
+                text: Constants.searchByFactionText,
+              ),
+              DrawerButton(
+                onClickAction: () {
+                  widget.hearthstoneCardListBloc
+                      .searchForCards(Endpoints.byCardSet + keyword);
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/search_menu.dart]: Same as above

File: lib/src/presentation/widget/welcome.dart [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+import 'package:hearthstoneapp/src/core/util/palette.dart';
+
+class Home extends StatelessWidget {
+  const Home({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Center(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(vertical: 60, horizontal: 20),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/welcome.dart]: Add values to const

File: lib/src/presentation/widget/welcome.dart [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        child: Column(
+          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+          children: [
+            Image.asset(Constants.hearthstoneFullLogo),
+            ElevatedButton(
+              style: ElevatedButton.styleFrom(
+                primary: Palette.boxColor,
+                padding: const EdgeInsets.all(10),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/welcome.dart]: Same as above

File: lib/src/presentation/widget/welcome.dart [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+              ),
+              onPressed: () {
+                Scaffold.of(context).openDrawer();
+              },
+              child: const Padding(
+                padding: EdgeInsets.all(8.0),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/welcome.dart]: Same as above

File: test/custom_matchers/has_state.dart [hunk 18]
```diff
@@ -0,0 +1,8 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:hearthstoneapp/src/core/util/api_response_state.dart';
+
+class HasState extends CustomMatcher {
+  HasState(matcher) : super('ApiResponse with state that is', 'state', matcher);
```
Comment (FranciscoGaraventa) [inline test/custom_matchers/has_state.dart]: trailing comma

File: test/unit_tests/fetch_cards_repository_test.dart [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+void main() {
+  group('Fetch cards', () {
+    test('Fetch cards searching by name', () async {
+      final datasource = MockApiService();
+      const endpoint = '${Endpoints.byName}ysera';
+
+      when(datasource.getData(endpoint: endpoint)).thenAnswer(
+        (_) async => http.Response(
+          '[{"cardId": "asd"}]',
+          200,
```
Comment (FranciscoGaraventa) [inline test/unit_tests/fetch_cards_repository_test.dart]: You can use HttpStatus.ok

File: test/unit_tests/fetch_cards_repository_test.dart [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+      );
+    });
+    test('If the code is not 200 it should return an empty list', () async {
+      final datasource = MockApiService();
+      const endpoint = '${Endpoints.byName}ysera';
+
+      when(datasource.getData(endpoint: endpoint)).thenAnswer(
+        (_) async => http.Response(
+          '[]',
+          404,
```
Comment (FranciscoGaraventa) [inline test/unit_tests/fetch_cards_repository_test.dart]: You can use HttpStatus.notFound

File: lib/src/data/repository/card_list_repository.dart (hunk 3 above)
Comment (FacuCristobalGlobant) [reply lib/src/data/repository/card_list_repository.dart:28]: I'm using a TextField to search for cards, so I'm returning an empty list as a way of saying that no cards were found that matched the criteria

File: lib/src/core/util/api_response_state.dart (hunk 1 above)
Comment (FacuCristobalGlobant) [reply lib/src/core/util/api_response_state.dart]: Added

File: lib/src/data/datasource/remote/api_service.dart (hunk 2 above)
Comment (FacuCristobalGlobant) [reply lib/src/data/datasource/remote/api_service.dart]: My IDE doesn't let me change to relative imports automatically using the context menu, so I was told it was ok leaving them like that.

File: lib/src/data/repository/card_list_repository.dart (hunk 3 above)
Comment (FacuCristobalGlobant) [reply lib/src/data/repository/card_list_repository.dart:28]: I added some logic to my bloc class to send an empty state event when receiving an empty list.

File: lib/src/domain/bloc/i_hearthstone_card_list_bloc.dart (hunk 4 above)
Comment (FacuCristobalGlobant) [reply lib/src/domain/bloc/i_hearthstone_card_list_bloc.dart:3]: Explained above.

File: lib/src/presentation/bloc/list_of_cards_bloc.dart (hunk 5 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/bloc/list_of_cards_bloc.dart:5]: Explained above.

File: lib/src/presentation/widget/card_property.dart (hunk 6 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/card_property.dart]: Explained above.

File: lib/src/presentation/widget/not_found.dart (hunk 7 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/not_found.dart]: Done

File: lib/src/presentation/widget/search_menu.dart (hunk 8 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done

File: lib/src/presentation/widget/search_menu.dart (hunk 9 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done

File: lib/src/presentation/widget/search_menu.dart (hunk 10 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done

File: lib/src/presentation/widget/search_menu.dart (hunk 11 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done

File: lib/src/presentation/widget/search_menu.dart (hunk 12 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done.

File: lib/src/presentation/widget/search_menu.dart (hunk 13 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done.

File: lib/src/presentation/widget/search_menu.dart (hunk 14 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/search_menu.dart]: Done.

File: lib/src/presentation/widget/welcome.dart (hunk 15 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/welcome.dart]: Done.

File: lib/src/presentation/widget/welcome.dart (hunk 16 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/welcome.dart]: Done.

File: lib/src/presentation/widget/welcome.dart (hunk 17 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/welcome.dart]: Done.

File: test/custom_matchers/has_state.dart (hunk 18 above)
Comment (FacuCristobalGlobant) [reply test/custom_matchers/has_state.dart]: Added.

File: test/unit_tests/fetch_cards_repository_test.dart (hunk 19 above)
Comment (FacuCristobalGlobant) [reply test/unit_tests/fetch_cards_repository_test.dart]: Done.

File: test/unit_tests/fetch_cards_repository_test.dart (hunk 20 above)
Comment (FacuCristobalGlobant) [reply test/unit_tests/fetch_cards_repository_test.dart]: Done.

File: lib/src/core/util/constants.dart [hunk 21]
```diff
@@ -3,16 +3,8 @@ import 'package:flutter/material.dart';
 abstract class Constants {
```
Comment (FranciscoGaraventa) [inline lib/src/core/util/constants.dart]: This constants file could be divided in different files, one relative to dimens (Like fontSize, height, Colors), the second one relative to String constants, another relative to api service calls (apiBaseUrl, api endpoints ..), and other only to assets.

File: lib/src/data/datasource/remote/api_service.dart [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+import 'package:http/http.dart' as http;
+
+class ApiService {
+  Future<http.Response> getData({
+    String endpoint = Constants.emptyString,
+  }) async {
+    final url = Uri.parse(
+      Constants.baseUrl + endpoint,
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/remote/api_service.dart]: Here you can made string interpolation

Comment (FacuCristobalGlobant) [reply lib/src/data/datasource/remote/api_service.dart]: Done.

File: lib/src/core/util/constants.dart (hunk 21 above)
Comment (FacuCristobalGlobant) [reply lib/src/core/util/constants.dart]: Done, I divided that file into Styles, Strings, ApiProperties and Assets.

**Review by LeonardoMolleker — APPROVED** [review]

File: lib/src/presentation/bloc/list_of_cards_bloc.dart [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+  Future<void> initialize() async {
+    _hearthstoneCardStreamController.sink.add(
+      _apiResponse,
+    );
+  }
+
+  @override
+  Stream<ApiResponse> getCardListStream() {
+    return _hearthstoneCardStreamController.stream;
+  }
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/list_of_cards_bloc.dart]: here instead to return like this way you can use a arrow function Stream<ApiResponse> getCardListStream => _heartstoneStreamController.stream;

**Review by FranciscoGaraventa — APPROVED** [review]
Approve with minor comment

File: lib/src/presentation/bloc/list_of_cards_bloc.dart (hunk 23 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/bloc/list_of_cards_bloc.dart]: Done.

**Review by MarcosCGlobant — APPROVED** [review]

**Pull request merged**